gravitee-node-kubernetes/src/main/java/io/gravitee/node/kubernetes/keystoreloader/KubernetesPemRegistryKeyStoreLoader.java

```diff
@@ -0,0 +1,197 @@
+
+        return Completable.merge(locationObs).andThen(Completable.fromRunnable(this::refreshKeyStoreBundle));
+    }
+
+    @Override
+    protected Flowable<ConfigMap> watch() {
+        final List<Flowable<Event<ConfigMap>>> toWatch = resources
+            .keySet()
+            .stream()
+            .map(location ->
+                kubernetesClient
+                    .watch(WatchQuery.<ConfigMap>from(location).build())
+                    .observeOn(Schedulers.computation())
+                    .repeat()
+                    .retryWhen(errors -> errors.delay(RETRY_DELAY_MILLIS, TimeUnit.MILLISECONDS))
+            )
+            .toList();
+
+        return Flowable.merge(toWatch).filter(event -> event.getType().equalsIgnoreCase("MODIFIED")).map(Event::getObject);
+    }
+
+    @Override
+    protected Completable loadKeyStore(ConfigMap configMap) {
+        if (options.getKeyStoreType().equalsIgnoreCase(CERTIFICATE_FORMAT_PEM_REGISTRY)) {
+            if (
+                configMap.getMetadata().getLabels() != null &&
+                CERTIFICATE_FORMAT_PEM_REGISTRY.equalsIgnoreCase(configMap.getMetadata().getLabels().get(GRAVITEEIO_PEM_REGISTRY_LABEL)) &&
+                configMap.getData() != null
+            ) {
+                return generateKeystoreFromPemRegistry(configMap);
+            } else {
+                // make sure there is no keys left from the past
+                keyStoresByLocation.put(GRAVITEEIO_PEM_REGISTRY_LABEL, initKeyStore());
+                return Completable.complete();
+            }
+        }
+
+        return Completable.error(new RuntimeException(String.format("unsupported keystore type %s", options.getKeyStoreType())));
+    }
+
+    private Completable generateKeystoreFromPemRegistry(ConfigMap configMap) {
+        return Maybe
+            .merge(
+                configMap
+                    .getData()
+                    .values()
+                    .stream()
+                    .flatMap(value -> Stream.of(value.replace("[", "").replace("]", "").replace("\"", "").trim().split(",")))
+                    .distinct()
+                    .filter(StringUtils::hasLength)
+                    .map(reference -> {
+                        String[] namespaceName = reference.split("/");
+                        if (namespaceName.length != 2) {
+                            return Maybe.<KeyStore>error(
+                                new IllegalArgumentException("Wrong or missing namespace or name of the TLS Secret")
+                            );
+                        }
+                        return kubernetesClient
+                            .get(ResourceQuery.secret(namespaceName[0], namespaceName[1]).build())
+                            .map(this::secretToKeyStore);
+                    })
+                    .toList()
+            )
+            .toList()
+            .map(keyStoreList -> KeyStoreUtils.merge(keyStoreList, options.getKeyStorePassword()))
```

Comment:
This is a kind of extra cost that we are paying for not watching every single secretes directly. If we use a map to avoid the merge, if a secret has been updated in the meantime, its new data will not be populated and will have inconsistency.

Reply:
fair enough